x/upgrade/abci_test.go

```diff
@@ -185,6 +209,162 @@ func (s *TestSuite) TestPlanStringer() {
+
+	for i = 1; i <= s.ctx.BlockHeight(); i++ {
+		newCtx = newCtx.WithBlockHeight(s.ctx.BlockHeight() + i)
+		s.Require().NotPanics(func() {
+			s.module.BeginBlock(newCtx, req)
+		})
+	}
+
+	s.T().Log("Verify a second proposal also is being cleared")
+	err = s.handler(s.ctx, SoftwareUpgradeProposal{Title: "prop2", Plan: Plan{Name: "test2", Height: s.ctx.BlockHeight() + 10}})
+	s.Require().Nil(err)
+	s.Require().NotPanics(func() {
+		s.module.BeginBlock(newCtx, req)
+	})
+
+	//To ensure verification is being done only after both uprades are cleared
+	s.T().Log("Verify if both proposals are cleared")
+	s.VerifyCleared(s.ctx)
+	s.VerifyNotDone(s.ctx, "test")
+	s.VerifyNotDone(s.ctx, "test2")
+}
+
+func (s *TestSuite) TestSkipUpgradeSkippingOne() {
+	newCtx := sdk.NewContext(s.cms, abci.Header{Height: s.ctx.BlockHeight() + 1, Time: time.Now()}, false, log.NewNopLogger())
+	req := abci.RequestBeginBlock{Header: newCtx.BlockHeader()}
+	err := s.handler(s.ctx, SoftwareUpgradeProposal{Title: "prop", Plan: Plan{Name: "test", Height: s.ctx.BlockHeight() + 1}})
+	s.Require().Nil(err)
+
+	var i int64
+
+	s.T().Log("Verify if skip upgrade flag clears upgrade plan in one case and does upgrade on another")
+	viper.Set(s.FlagUnsafeSkipUpgrade, []int{int(s.ctx.BlockHeight() + 1)})
+	s.VerifySet()
+
+	s.VerifyConversion(viper.GetIntSlice(s.FlagUnsafeSkipUpgrade))
+
+	for i = 1; i < s.ctx.BlockHeight(); i++ {
+		//To execute begin block multiple times with a new block height
+		newCtx = newCtx.WithBlockHeight(s.ctx.BlockHeight() + i)
+		s.Require().NotPanics(func() {
+			s.module.BeginBlock(newCtx, req)
+		})
+	}
+
+	s.T().Log("Verify the second proposal is not skipped")
+	err = s.handler(s.ctx, SoftwareUpgradeProposal{Title: "prop2", Plan: Plan{Name: "test2", Height: s.ctx.BlockHeight() + 10}})
+	s.Require().Nil(err)
+	newCtx = newCtx.WithBlockHeight(newCtx.BlockHeight() + 1)
+	s.VerifyDoUpgradeWithCtx(newCtx, "test2")
+
+	s.T().Log("Verify first proposal is cleared and second is done")
+	s.VerifyNotDone(s.ctx, "test")
+	s.VerifyDone(s.ctx, "test2")
+}
+
+func (s *TestSuite) TestSkipUpgradeIgnoringBoth() {
+	newCtx := sdk.NewContext(s.cms, abci.Header{Height: s.ctx.BlockHeight() + 1, Time: time.Now()}, false, log.NewNopLogger())
+	req := abci.RequestBeginBlock{Header: newCtx.BlockHeader()}
+	err := s.handler(s.ctx, SoftwareUpgradeProposal{Title: "prop", Plan: Plan{Name: "test", Height: s.ctx.BlockHeight() + 1}})
+	s.Require().Nil(err)
+
+	var i int64
+
+	s.T().Log("Verify if skip upgrade flag clears upgrade plan in both cases and does third upgrade")
+	viper.Set(s.FlagUnsafeSkipUpgrade, []int{int(s.ctx.BlockHeight() + 1), int(s.ctx.BlockHeight() + 10)})
+	s.VerifySet()
+
+	s.VerifyConversion(viper.GetIntSlice(s.FlagUnsafeSkipUpgrade))
+
+	for i = 1; i < s.ctx.BlockHeight(); i++ {
+		newCtx = newCtx.WithBlockHeight(s.ctx.BlockHeight() + i)
+		s.Require().NotPanics(func() {
+			s.module.BeginBlock(newCtx, req)
+		})
+	}
+
+	//A new proposal with height in skip
```

Comment:
Same as above, both for lop, and `newCtx = newCtx.WithBlockHeight(newCtx.BlockHeight() + 1) ` below